transient_value.h compilation error

[100%] Building CXX object test/type/CMakeFiles/transient_value_test.dir/transient_value_test.cpp.o
In file included from /home/pakhtar/CLionProjects/catalog/test/type/transient_value_test.cpp:1:0:
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = bool]’:
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value_peeker.h:24:30:   required from here
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h:249:12: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
     return *reinterpret_cast<const T *const>(&data_);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = short int]’:
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value_peeker.h:46:33:   required from here
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h:249:12: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = int]’:
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value_peeker.h:57:33:   required from here
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h:249:12: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = double]’:
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value_peeker.h:79:32:   required from here
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h:249:12: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = terrier::common::StrongTypeAlias<terrier::type::tags::date_t_typedef_tag, unsigned int>]’:
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value_peeker.h:101:32:   required from here
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h:249:12: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h:249:52: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
     return *reinterpret_cast<const T *const>(&data_);
                                                    ^
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = const char* const]’:
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value_peeker.h:113:64:   required from here
/home/pakhtar/CLionProjects/catalog/src/include/type/transient_value.h:249:12: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
     return *reinterpret_cast<const T *const>(&data_);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

From when the transient value PR was merged. The same issue is in the logs as a warning, so the builds were not flagged as failures. 

[ 66%] Building CXX object test/storage/CMakeFiles/garbage_collector_test.dir/garbage_collector_test.cpp.o

In file included from /jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/parser/expression/abstract_expression.h:10:0,

                 from /jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/parser/expression/comparison_expression.h:6,

                 from /jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/test/parser/parser_test.cpp:6:

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = bool]’:

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value_peeker.h:24:30:   required from here

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h:249:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]

     return *reinterpret_cast<const T *const>(&data_);

            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = short int]’:

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value_peeker.h:46:33:   required from here

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h:249:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = int]’:

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value_peeker.h:57:33:   required from here

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h:249:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = double]’:

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value_peeker.h:79:32:   required from here

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h:249:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = terrier::common::StrongTypeAlias<terrier::type::tags::date_t_typedef_tag, unsigned int>]’:

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value_peeker.h:101:32:   required from here

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h:249:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h:249:52: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]

     return *reinterpret_cast<const T *const>(&data_);

                                                    ^

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h: In instantiation of ‘T terrier::type::TransientValue::GetAs() const [with T = const char* const]’:

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value_peeker.h:113:64:   required from here

/jenkins/workspace/terrier_PR-260-H5ZSJJQ7Y76FIFTWMAITVEKE7MWPNNEEQYEM2FAK455T5JSDKY6A/src/include/type/transient_value.h:249:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]

     return *reinterpret_cast<const T *const>(&data_);

            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

--- a/src/include/type/transient_value.h
+++ b/src/include/type/transient_value.h
@@ -246,7 +246,8 @@ class TransientValue {
    */
   template <typename T>
   T GetAs() const {
-    return *reinterpret_cast<const T *const>(&data_);
+    const auto *const value = reinterpret_cast<const T *const>(&data_);
+    return *value;
   }
 
   /**

